hawc/apps/common/views.py

```diff
@@ -418,27 +419,45 @@ def get_template_names(self):
         return [name]
 
 
+class WebappMixin:
+    """Mixin to startup a javascript single-page application"""
+
+    get_app_config: Optional[Callable[[RequestContext], WebappConfig]] = None
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        if self.get_app_config:
+            context["config"] = self.get_app_config(context).dict()
+        return context
+
+
 # Base HAWC views
 class BaseDetail(AssessmentPermissionsMixin, DetailView):
     crud = "Read"
-    breadcrumb_active_name: Optional[str] = None  # optional
+    breadcrumb_active_name: Optional[str] = None
+    get_app_config: Optional[Callable[[RequestContext], WebappConfig]] = None
 
     def get_breadcrumbs(self) -> List[Breadcrumb]:
         return Breadcrumb.build_assessment_crumbs(self.request.user, self.object)
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
-        context["assessment"] = self.assessment
-        context["crud"] = self.crud
-        context["obj_perms"] = super().get_obj_perms()
-        context["breadcrumbs"] = self.get_breadcrumbs()
+        context.update(
```

Comment:
ahhh, ok now i'm with you. I did this originally actually, or something similar.  The problem I hit was the class method resolution hierarchy. I was finding a case where with a few classes, but not all, assessment wasnt yet available b/c the webapp mixin was called first. 

https://github.com/shapiromatron/hawc/blob/83d0e7ae1e570e03f5a15e9cd551fb088a152f04/hawc/apps/common/views.py#L420-L429

I didnt explore it too deep b/c it seemed like it would vary on each view, but maybe I should go back? Instead o putting this mixin at the far-left, perhaps it goes to to the far-right except for the base class?

Reply:
The ordering of the parent classes don't help here, since the additional context keys (`assessment`, etc) that are sometimes needed by `WebappMixin` are added in the subclass (`BaseDetail`, etc). So the only workaround I can think of is to add those to the context first, _then_ call the method from `super`. I added a commit that does that, but it only got rid of 11 lines of code, so feel free to revert it if you're not a fan of the pattern.